### PR TITLE
Api error formatting

### LINKED ISSUE #168: Implement API guidelines (validation) error format
The default DRF validation error format does not map to the format used by Dutch APIs. We should make sure we do that.

For inspiration: https://github.com/VNG-Realisatie/vng-api-common/blob/master/vng_api_common/exception_handling.py#L67

### PR DESCRIPTION
**Pull request by shea-maykinmedia**
Fixes #168 

Most of this code was taken from `vng_api_common` and the tests within `exception_handler` were taken from open zaak. 

**Screenshots**

Screenshot 2021-07-28 at 12 20 33

Screenshot 2021-07-28 at 12 20 47

### CODE REVIEW

File: src/openforms/exception_handler/exceptions.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+class Gone(APIException):
+    status_code = status.HTTP_410_GONE
+    default_detail = _("The resource is gone")
+    default_code = "gone"
+
+
+class PreconditionFailed(APIException):
+    status_code = status.HTTP_412_PRECONDITION_FAILED
+    default_detail = _("Precondition failed")
+    default_code = "precondition_failed"
```
Comment (sergei-maertens) [inline src/openforms/exception_handler/exceptions.py:22]: This belongs in `openforms.api.exceptions` to follow the DRF schema (`rest_framework.exceptions`)

File: src/openforms/exception_handler/serializers.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
+from rest_framework import serializers
+
+
+class FieldValidationErrorSerializer(serializers.Serializer):
+    """
+    Formaat van validatiefouten.
+    """
```
Comment (sergei-maertens) [inline src/openforms/exception_handler/serializers.py:7]:
```suggestion
    """
    Validation error format, following the NL API Strategy.
    
    See https://docs.geostandaarden.nl/api/API-Strategie/ and 
    https://docs.geostandaarden.nl/api/API-Strategie-ext/#error-handling-0
    """
```

File: src/openforms/exception_handler/serializers.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    code = serializers.CharField(help_text="Systeemcode die het type fout aangeeft")
+    reason = serializers.CharField(
+        help_text="Uitleg wat er precies fout is met de gegevens"
+    )
+
+
+class FoutSerializer(serializers.Serializer):
+    """
+    Formaat van HTTP 4xx en 5xx fouten.
+    """
```
Comment (sergei-maertens) [inline src/openforms/exception_handler/serializers.py:19]:
```suggestion
    """
    Error format for HTTP 4xx and 5xx errors.
    
    See https://docs.geostandaarden.nl/api/API-Strategie-ext/#error-handling-0 for the NL API strategy guidelines.
    """
```

File: src/openforms/exception_handler/serializers.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+    type = serializers.CharField(
+        help_text="URI referentie naar het type fout, bedoeld voor developers",
+        required=False,
+        allow_blank=True,
+    )
+    # not according to DSO, but possible for programmatic checking
```
Comment (sergei-maertens) [inline src/openforms/exception_handler/serializers.py:26]:
```suggestion
```

File: src/openforms/exception_handler/serializers.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    code = serializers.CharField(help_text="Systeemcode die het type fout aangeeft")
+    title = serializers.CharField(help_text="Generieke titel voor het type fout")
+    status = serializers.IntegerField(help_text="De HTTP status code")
+    detail = serializers.CharField(
+        help_text="Extra informatie bij de fout, indien beschikbaar"
+    )
+    instance = serializers.CharField(
+        help_text="URI met referentie naar dit specifiek voorkomen van de fout. Deze kan "
+        "gebruikt worden in combinatie met server logs, bijvoorbeeld."
+    )
```
Comment (sergei-maertens) [inline src/openforms/exception_handler/serializers.py:36]: Use translatable `help_text` messages, in English.

File: src/openforms/exception_handler/urls.py [hunk 6]
```diff
@@ -0,0 +1,9 @@
+from django.urls import path
+
+from .views import ErrorDetailView
+
+app_name = "exception_handler"
+
+urlpatterns = [
+    path("fouten/<exception_class>/", ErrorDetailView.as_view(), name="error-detail"),
+]
```
Comment (sergei-maertens) [inline src/openforms/exception_handler/urls.py:9]: Please put all of the code in this package/app into the `openforms.api` package.

The URL route can be included directly from `openforms.urls` so we don't get the `/api` prefix for the human-facing views.

File: src/openforms/exception_handler/utils.py [hunk 7]
```diff
@@ -0,0 +1,18 @@
+import re
+from typing import Union
+
+from djangorestframework_camel_case.util import (
+    underscore_to_camel as _underscore_to_camel,
+)
+
+RE_UNDERSCORE = re.compile(r"[a-z]_[a-z]")
```
Comment (sergei-maertens) [inline src/openforms/exception_handler/utils.py:8]: Use `djangorestframework_camel_case.util.camelize_re` instead.

File: src/openforms/forms/tests/test_api.py [hunk 8]
```diff
@@ -157,8 +162,24 @@ def test_create_form_unsuccessful_with_bad_data(self):
         self.assertEqual(
             response.json(),
             {
-                "name": [_("This field is required.")],
-                "slug": [_("This field is required.")],
+                "type": "http://testserver/exception-handler/fouten/ValidationError/",
+                "code": "invalid",
+                "title": "Invalid input.",
```
Comment (sergei-maertens) [inline src/openforms/forms/tests/test_api.py:167]: These should also be marked as translatable and go through the translation machinery.

**Review by sergei-maertens — CHANGES_REQUESTED** [review]

File: src/openforms/api/tests/error_views.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+        "application/xml", detail="This media type is not supported"
+    )
+
+
+class ThrottledView(BaseErrorView):
+    exception = exceptions.Throttled(detail="Too many requests")
+
+
+class InternalServerErrorView(BaseErrorView):
+    exception = exceptions.APIException("Everything broke")
```
Comment (sergei-maertens) [inline src/openforms/api/tests/error_views.py:68]: Please ensure these strings are marked translatable too.

File: src/openforms/api/tests/test_error_format.py [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+                ],
+            },
+        )
+
+    def test_401_error(self):
+        self.assertErrorResponse(
+            views.NotAuthenticatedView,
+            {
+                "code": "not_authenticated",
+                "title": "Authenticatiegegevens zijn niet opgegeven.",
```
Comment (sergei-maertens) [inline src/openforms/api/tests/test_error_format.py:58]: Please ensure _every_ test uses translatable strings _everywhere_ for the user-facing messages. I'm not going to copy paste the same comment for every individual test.

File: src/openforms/api/views.py [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+        exc_klass = self._get_exception_klass()
+        context.update(
+            {
+                "type": exc_klass.__name__,
+                "status_code": exc_klass.status_code,
+                "default_detail": exc_klass.default_detail,
+                "default_code": exc_klass.default_code,
+            }
+        )
+        return context
```
Comment (sergei-maertens) [inline src/openforms/api/views.py:69]: Please move this to `openforms.utils.views` - mixing regular django views and DRF views is confusing.

File: src/openforms/conf/base.py [hunk 12]
```diff
@@ -596,6 +596,7 @@
     "DEFAULT_THROTTLE_RATES": {"anon": "200/day", "user": "1000/day"},
     "TEST_REQUEST_DEFAULT_FORMAT": "json",
     "DEFAULT_SCHEMA_CLASS": "openforms.api.schema.AutoSchema",
+    "EXCEPTION_HANDLER": "openforms.api.views.handler",
```
Comment (sergei-maertens) [inline src/openforms/conf/base.py:599]: `handler` really says nothing - and you should really follow the upstream pattern. We're overriding the default DRF exception handler: https://github.com/encode/django-rest-framework/blob/master/rest_framework/views.py#L71

Please use the same function name.

**Review by sergei-maertens — CHANGES_REQUESTED** [review]

**Comment by shea-maykinmedia** [comment]
I double checked all the files I believe I marked all the necessary strings as translatable so hopefully I didn't miss any :)

**Review by sergei-maertens — APPROVED** [review]

**Pull request merged**